tools/erc-repository-indexer/erc-contract-indexer/src/services/byteCodeAnalyzer.ts

```diff
@@ -0,0 +1,111 @@
+ * limitations under the License.
+ *
+ */
+
+import { Contract } from 'sevm';
+import { ContractScannerService } from './contractScanner';
+import { MirrorNodeContract } from '../schemas/MirrorNodeSchemas';
+import { ERCOutputInterface } from '../schemas/ERCRegistrySchemas';
+
+export enum ERCID {
+  ERC20 = 'ERC20',
+  ERC721 = 'ERC721',
+}
+
+export class ByteCodeAnalyzer {
+  /**
+   * Analyzes bytecode, detects and categorizes contracts into ERC20 and ERC721 types based on their bytecode.
+   * @param {ContractScannerService} contractScannerService - The service used to fetch contract bytecode.
+   * @param {MirrorNodeContract[]} contractObject - An array of contract objects to categorize.
+   * @returns {Promise<{erc20Contracts: ERCOutputInterface[], erc721Contracts: ERCOutputInterface[]}>} An object containing arrays of categorized ERC20 and ERC721 contracts.
+   * @throws {Error} If there's an error while analyzing contract bytecode.
+   */
+  async categorizeERCContracts(
+    contractScannerService: ContractScannerService,
+    contractObject: MirrorNodeContract[]
+  ): Promise<{
+    erc20Contracts: ERCOutputInterface[];
+    erc721Contracts: ERCOutputInterface[];
+  }> {
+    const erc20Contracts: ERCOutputInterface[] = [];
+    const erc721Contracts: ERCOutputInterface[] = [];
+
+    try {
+      const contractResponses = await Promise.all(
+        contractObject.map(({ contract_id }) =>
+          contract_id
+            ? contractScannerService.fetchContractByteCode(contract_id)
+            : null
+        )
+      );
+
+      contractResponses.forEach((contract) => {
+        if (
+          !contract ||
+          !contract.bytecode ||
+          !contract.contract_id ||
+          !contract.evm_address ||
+          !contract.runtime_bytecode
+        ) {
+          console.warn('Skipping contract due to missing data:', {
+            contractId: contract?.contract_id,
+            hasBytecode: !!contract?.bytecode,
+            hasContractId: !!contract?.contract_id,
+            hasEvmAddress: !!contract?.evm_address,
+            hasRuntimeBytecode: !!contract?.runtime_bytecode,
+          });
+          return;
+        }
+        const contractBytecode =
+          contract.runtime_bytecode === '0x'
+            ? contract.bytecode
+            : contract.runtime_bytecode;
+
+        console.log(`Analyzing contract: contractId=${contract.contract_id}`);
+
+        const sevmContract = new Contract(contractBytecode);
+        const ercOutput: ERCOutputInterface = {
+          address: contract.evm_address,
+          contractId: contract.contract_id,
+        };
+
+        if (sevmContract.isERC(ERCID.ERC20)) {
+          console.log(
+            `New ERC contract detected: contractId=${contract.contract_id}, ercID: ${ERCID.ERC20}`
+          );
+          erc20Contracts.push(ercOutput);
+
+          // TODO: Make calls to MN to retrieve name, symbol, decimals, totalSuply, etc.
```

Comment:
Great point thanks for pointing it out! 

Yes, that was the the plan. All independent MN requests will be submitted concurrently to avoid slowing down performance. Just how it was done for the bytecode fetching above.